core/src/apps/cardano/shelley/address.py

```diff
@@ -0,0 +1,137 @@
+from trezor import wire
+from trezor.crypto import hashlib
+from trezor.messages import CardanoAddressType
+
+from apps.cardano.shelley import CURVE
+from apps.cardano.shelley.bech32 import bech32_encode
+from apps.cardano.shelley.bootstrap_address import derive_address_and_node
+from apps.cardano.shelley.utils import variable_length_encode
+from apps.common import HARDENED, paths
+from apps.common.seed import remove_ed25519_prefix
+
+
+def validate_full_path(path: list) -> bool:
+    """
+    Validates derivation path to fit {44', 1852'}/1815'/a'/{0,1}/i,
+    where `a` is an account number and i an address index.
+    The max value for `a` is 20, 1 000 000 for `i`.
+    """
+    if len(path) != 5:
+        return False
+    if path[0] != 44 | HARDENED and path[0] != 1852 | HARDENED:
+        return False
+    if path[1] != 1815 | HARDENED:
+        return False
+    if (
+        path[2] < HARDENED or path[2] > 20 | HARDENED
+    ):  # TODO do we still limit it to 20 accounts?
+        return False
+    if path[3] != 0 and path[3] != 1:
+        return False
+    if path[4] > 1000000:  # TODO do we still impose this?
+        return False
+    return True
+
+
+async def validate_address_path(ctx, msg, keychain):
+    # TODO what does this do? should we call it somewhere for validation of staking key paths, too?
+    await paths.validate_path(ctx, validate_full_path, keychain, msg.address_n, CURVE)
+
+
+def get_human_readable_address(address):
+    return bech32_encode("addr", address)
+
+
+def get_base_address(keychain, path: list, network_id):
+    if not _validate_shelley_address_path(path):
+        raise wire.DataError("Invalid path for base address")
+
+    spending_part = _get_spending_part(keychain, path)
+
+    staking_path = _path_to_staking_path(path)
+    staking_node = keychain.derive(staking_path)
+    staking_key = remove_ed25519_prefix(staking_node.public_key())
+    staking_part = hashlib.blake2b(data=staking_key, outlen=28).digest()
+
+    address_header = _get_address_header(CardanoAddressType.BASE_ADDRESS, network_id)
+    address = address_header + spending_part + staking_part
+
+    return get_human_readable_address(address)
+
+
+def get_pointer_address(
+    keychain, path: list, network_id, block_index, tx_index, certificate_index
```

Comment:
type annotations missing

Reply:
Yes, as I mentioned above - I will deal with all type annotations when I've joined the byron/shelley folderes.